platform/src/server/http/http_server.ts

```diff
@@ -55,41 +86,79 @@ export class HttpServer {
-    // TODO This was forced to `any` because of the following problem with the
-    // express declaration file, and can be removed once that is fixed:
-    //
-    // ```
-    // Argument of type 'Application' is not assignable to parameter of type
-    // '((request: IncomingMessage, response: ServerResponse) => void) | undefined'
-    // ```
-    const app = this.app as any;
-
-    if (!config.ssl.enabled) {
-      return http.createServer(app);
-    }
+    const options: ServerOptions = {
+      host: config.host,
+      port: config.port,
+      routes: {
+        cors: config.cors,
+        payload: {
+          maxBytes: config.maxPayload.getValueInBytes(),
+        },
+        validate: {
+          options: {
+            abortEarly: false
+          }
+        }
+      },
+      state: {
+        strictHeader: false
+      },
+    };
+
+    const ssl = config.ssl;
+    if (ssl.enabled) {
+      const tlsOptions: TLSOptions = {
+        ca: config.ssl.certificateAuthorities &&
+          config.ssl.certificateAuthorities.map(caFilePath => readFileSync(caFilePath)),
+
+        key: readFileSync(ssl.key!),
+        cert: readFileSync(ssl.certificate!),
+        passphrase: ssl.keyPassphrase,
 
-    // TODO: add support for `secureOptions`.
-    return https.createServer(
-      {
-        key: readFileSync(config.ssl.key!),
-        cert: readFileSync(config.ssl.certificate!),
-        ca:
-          config.ssl.certificateAuthorities &&
-          config.ssl.certificateAuthorities.map(caFilePath =>
-            readFileSync(caFilePath)
-          ),
-        passphrase: config.ssl.keyPassphrase,
         ciphers: config.ssl.cipherSuites.join(':'),
         // We use the server's cipher order rather than the client's to prevent the BEAST attack.
         honorCipherOrder: true,
-      },
-      app
-    );
+        secureOptions: ssl.getSecureOptions(),
+      };
+
+      // TODO: Hapi types have a typo in `tls` property type definition: `https.RequestOptions` is used instead of
+      // `https.ServerOptions`, and `honorCipherOrder` isn't presented in `https.RequestOptions`.
+      options.tls = tlsOptions as any;
+    }
+
+    return new Server(options);
+  }
+
+  private setupBasePathRewrite(server: Server, config: HttpConfig) {
+    if (config.basePath === undefined || !config.rewriteBasePath) {
+      return;
+    }
+
+    const basePath = config.basePath;
+    server.ext('onRequest', (request: Request, h: ResponseToolkit) => {
+      const newURL = modifyUrl(request.url.href!, (urlParts) => {
+        if (urlParts.pathname != null && urlParts.pathname.startsWith(basePath)) {
+          urlParts.pathname = urlParts.pathname.replace(basePath, '') || '/';
+        } else {
+          return {};
+        }
+      });
+
+      if (!newURL) {
+        return h.response('Not Found').code(404);
+      }
+
+      request.setUrl(newURL);
+      // Raw request can be forwarded to the old platform that doesn't expect base path
```

Comment:
This comment is confusing. Should this be actually two sentences? I feel a split between "platform" and "that".

Reply:
Yeah, agree, will something like this be better: `We should update raw request as well since it can be proxied to the old platform where base path isn't expected.`?

Reply:
:heavy_check_mark: (changed to `We should update raw request as well since it can be proxied to the old platform where base path isn't expected.`, feel free to propose a better alternative - I'll fix that.)